test/server/http/admin_test.cc

```diff
@@ -77,17 +82,42 @@ class AdminInstanceTest : public testing::TestWithParam<Network::Address::IpVers
         cpu_profile_path_(TestEnvironment::temporaryPath("envoy.prof")),
         admin_("/dev/null", cpu_profile_path_, address_out_path_,
                Network::Test::getCanonicalLoopbackAddress(GetParam()), server_,
-               listener_scope_.createScope("listener.admin.")) {
+               listener_scope_.createScope("listener.admin.")),
+        expect_no_logs_(true) {
+
     EXPECT_EQ(std::chrono::milliseconds(100), admin_.drainTimeout());
     admin_.tracingStats().random_sampling_.inc();
     EXPECT_TRUE(admin_.setCurrentClientCertDetails().empty());
+    Logger::Registry::setLogLevel(Logger::Logger::warn);
```

Comment:
I commented on this in a recent @zuercher PR, but I don't think this plays well with when someone sets the log level on the CLI for a test (via `--test_arg`). a) it will overwrite what they set, which I guess might be OK, but b) I think it will leave it set for all tests in the same binary. Should this level be somehow stackable also such that you could potentially just set up your delegate with level and then have everything be restored? (logger/level/format)

Reply:
On other option, as food for thought: I believe spdlog supports multiple sinks. If not, instead of a delegation system we could actually allow a list system, with independent formatting/level. This might be a safer way of doing this type of testing. Basically, just add the mock logger to the end of the list of already existing loggers.

Reply:
Ah, good catch about the test resetting. I had glanced at test_runner.h, which resets the logging system, but now I see it only does it once for the binary, not once per testcase.  I was thinking that was the hook you had mentioned to me about totally resetting all singleton state between each test. Now I'm not sure where such a between-test-global-reset is supposed to occur. I think this issue is somewhat orthogonal to mocking the logs, but I can add a similar RAII mechanism to this PR, which can restore the config mechanisms on destruction.  I think this can be independent.

spdlog does support mulitple sinks in 2 ways, from my re-reading:
1. You can use a dist_sink, which contains a vector<sink> and exposes an add/remove of individual sinks.  Note that if we use this, we'll take an extra mutex in the spdlog system on each
log or flush, beyond what we have in our code.
2. You can specify a list of sinks up front when you create the logger, which I don't think helps us.

I had seen the list-of-sinks-up-front option when designing this solution, but missed the dist_sink, which could be made to work.

However, I still slightly prefer the RAII solution I have here because:
  - I kind of think the tests shouldn't be noisy, and the mocked logging should replace the actual logging so the test output is not noisy. OTOH that makes the actual logging code path untested by code that uses it; of course it would still be tested by its own tests.
  - the RAII approach to stacking the mock feels right to me, and that's annoying to do with
    spdylog sinks which are all shared_ptr.

If that's OK with you, then I'll work on adding some more comments.

Reply:
> I kind of think the tests shouldn't be noisy, and the mocked logging should replace the actual logging so the test output is not noisy. OTOH that makes the actual logging code path untested by code that uses it; of course it would still be tested by its own tests.

The issue is that sometimes you need the logging to debug the test. ;)

I agre in this trivial example it doesn't matter that much, but what about more complex examples?

In general I don't have any objection to this approach, I just want to make sure that it doesn't regress dev/test/debug flow. Thoughts?